File: bench/algorithm/regex-redux/5.go
/* The Computer Language Benchmarks Game
 * https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 *
 * regex-dna program contributed by The Go Authors.
 * modified by Tylor Arndt.
 * modified by Chandra Sekar S to use optimized PCRE binding.
 * modified by Matt Dellandrea.
 * modified by Pavel Griaznov to use PCRE JIT compilation.
 */

package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"runtime"

	"github.com/GRbit/go-pcre"
)

type substitution struct {
	pattern     string
	replacement string
}

func countMatches(pat string, b []byte) int {
	m := pcre.MustCompileJIT(pat, 0, pcre.STUDY_JIT_COMPILE).Matcher(b, 0)
	n := 0

	for f := m.Matches; f; f = m.Match(b, 0) {
		n++

		b = b[m.Index()[1]:]
	}

	return n
}

func main() {
	fileName := "25000_in"
	if len(os.Args) > 1 {
		fileName = os.Args[1]
	}

	var variants = []string{
		"agggtaaa|tttaccct",
		"[cgt]gggtaaa|tttaccc[acg]",
		"a[act]ggtaaa|tttacc[agt]t",
		"ag[act]gtaaa|tttac[agt]ct",
		"agg[act]taaa|ttta[agt]cct",
		"aggg[acg]aaa|ttt[cgt]ccct",
		"agggt[cgt]aa|tt[acg]accct",
		"agggta[cgt]a|t[acg]taccct",
		"agggtaa[cgt]|[acg]ttaccct",
	}

	var substs = []substitution{
		{"tHa[Nt]", "<4>"},
		{"aND|caN|Ha[DS]|WaS", "<3>"},
		{"a[NSt]|BY", "<2>"},
		{"<[^>]*>", "|"},
		{"\\|[^|][^|]*\\|", "-"},
	}

	runtime.GOMAXPROCS(runtime.NumCPU())

	if f, err := os.Open(fileName); err == nil {
		defer f.Close()
		b, err := ioutil.ReadAll(f)
		if err != nil {
			fmt.Fprintf(os.Stderr, "can't read input: %s\n", err)
			os.Exit(2)
		}

		ilen := len(b)

		// Delete the comment lines and newlines
		b = pcre.
			MustCompileJIT("(>[^\n]*)?\n", 0, pcre.STUDY_JIT_COMPILE).
			ReplaceAll(b, []byte{}, 0)
		clen := len(b)

		mresults := make([]chan int, len(variants))
		for i := 0; i < len(variants); i++ {
			mresults[i] = make(chan int)

			go func(ch chan int, s string) {
				ch <- countMatches(s, b)
			}(mresults[i], variants[i])
		}

		lenresult := make(chan int)

		go func(b []byte) {
			for i := 0; i < len(substs); i++ {
				b = pcre.
					MustCompileJIT(substs[i].pattern, 0, pcre.STUDY_JIT_COMPILE).
					ReplaceAll(b, []byte(substs[i].replacement), 0)
			}
			lenresult <- len(b)
		}(b)

		for i := 0; i < len(variants); i++ {
			fmt.Printf("%s %d\n", variants[i], <-mresults[i])
		}

		fmt.Printf("\n%d\n%d\n%d\n", ilen, clen, <-lenresult)
	} else {
		panic(err)
	}
}

File: bench/algorithm/regex-redux/5.cs
/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 
   Regex-Redux by Josh Goldfoot
   order variants by execution time by Anthony Lloyd
*/

using System;
using System.IO;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

public static partial class RegexRedux
{
    public static async Task Main(string[] args)
    {
        var fileName = args.Length > 0 ? args[0] : "25000_in";
        var sequences = await File.ReadAllTextAsync(fileName).ConfigureAwait(false);
        var initialLength = sequences.Length;
        sequences = Regex.Replace(sequences, ">.*\n|\n", "");

        var magicTask = () =>
        {
            var newseq = Regex.Replace(sequences, "tHa[Nt]", "<4>", RegexOptions.Compiled);
            newseq = Regex.Replace(newseq, "aND|caN|Ha[DS]|WaS", "<3>", RegexOptions.Compiled);
            newseq = Regex.Replace(newseq, "a[NSt]|BY", "<2>", RegexOptions.Compiled);
            newseq = Regex.Replace(newseq, "<[^>]*>", "|", RegexOptions.Compiled);
            newseq = Regex.Replace(newseq, "\\|[^|][^|]*\\|", "-", RegexOptions.Compiled);
            return newseq.Length;
        };

        var variant2 = "[cgt]gggtaaa|tttaccc[acg] " + Regex.Count(sequences, "[cgt]gggtaaa|tttaccc[acg]", RegexOptions.Compiled);
        var variant3 = "a[act]ggtaaa|tttacc[agt]t " + Regex.Count(sequences, "a[act]ggtaaa|tttacc[agt]t", RegexOptions.Compiled);
        var variant7 = "agggt[cgt]aa|tt[acg]accct " + Regex.Count(sequences, "agggt[cgt]aa|tt[acg]accct", RegexOptions.Compiled);
        var variant6 = "aggg[acg]aaa|ttt[cgt]ccct " + Regex.Count(sequences, "aggg[acg]aaa|ttt[cgt]ccct", RegexOptions.Compiled);
        var variant4 = "ag[act]gtaaa|tttac[agt]ct " + Regex.Count(sequences, "ag[act]gtaaa|tttac[agt]ct", RegexOptions.Compiled);
        var variant5 = "agg[act]taaa|ttta[agt]cct " + Regex.Count(sequences, "agg[act]taaa|ttta[agt]cct", RegexOptions.Compiled);
        var variant1 = "agggtaaa|tttaccct " + Regex.Count(sequences, "agggtaaa|tttaccct", RegexOptions.Compiled);
        var variant9 = "agggtaa[cgt]|[acg]ttaccct " + Regex.Count(sequences, "agggtaa[cgt]|[acg]ttaccct", RegexOptions.Compiled);
        var variant8 = "agggta[cgt]a|t[acg]taccct " + Regex.Count(sequences, "agggta[cgt]a|t[acg]taccct", RegexOptions.Compiled);

        Console.WriteLine(variant1());
        Console.WriteLine(variant2());
        Console.WriteLine(variant3());
        Console.WriteLine(variant4());
        Console.WriteLine(variant5());
        Console.WriteLine(variant6());
        Console.WriteLine(variant7());
        Console.WriteLine(variant8());
        Console.WriteLine(variant9());
        Console.WriteLine($"\n{initialLength}\n{sequences.Length}");
        Console.WriteLine(magicTask().ToString());
    }
}
